tests/thread_tests/remote.go

```diff
@@ -0,0 +1,209 @@
+		remoteSnapPathEnv       = "REMOTE_SNAP_PATH"
+	)
+
+	if v := os.Getenv(remoteUserEnv); v != "" {
+		remoteUser = v
+	}
+
+	if v := os.Getenv(remotePasswordEnv); v != "" {
+		remotePassword = v
+	}
+
+	if v := os.Getenv(remoteHostEnv); v != "" {
+		remoteHost = v
+	}
+
+	if v := os.Getenv(remoteInfraInterfaceEnv); v != "" {
+		remoteInfraInterface = v
+	}
+
+	if v := os.Getenv(remoteSnapPathEnv); v != "" {
+		remoteSnapPath = v
+	}
+}
+
+func connectSSH(t *testing.T) {
+	if SSHClient != nil {
+		return
+	}
+
+	config := &ssh.ClientConfig{
+		User: remoteUser,
+		Auth: []ssh.AuthMethod{
+			ssh.Password(remotePassword),
+		},
+		Timeout:         10 * time.Second,
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+
+	var err error
+	SSHClient, err = ssh.Dial("tcp", remoteHost+":22", config)
+	if err != nil {
+		t.Fatalf("Failed to dial: %s", err)
+	}
+
+	t.Cleanup(func() {
+		SSHClient.Close()
+	})
+
+	t.Logf("SSH: connected to %s", remoteHost)
+}
+
+func remote_deployOTBRAgent(t *testing.T) {
+
+	t.Cleanup(func() {
+		remote_exec(t, "sudo snap remove --purge openthread-border-router")
+	})
+
+	start := time.Now().UTC()
+
+	commands := []string{
+		"sudo snap remove --purge openthread-border-router",
+		"sudo snap install openthread-border-router --edge",
+		"sudo snap set openthread-border-router infra-if='" + remoteInfraInterface + "'",
+		"sudo snap set openthread-border-router webgui-port=5000",
+		// "sudo snap connect openthread-border-router:avahi-control",
+		"sudo snap connect openthread-border-router:firewall-control",
+		"sudo snap connect openthread-border-router:raw-usb",
+		"sudo snap connect openthread-border-router:network-control",
+		// "sudo snap connect openthread-border-router:bluetooth-control",
+		// "sudo snap connect openthread-border-router:bluez",
+		"sudo snap start openthread-border-router",
+	}
+	for _, cmd := range commands {
+		remote_exec(t, cmd)
+	}
+
+	remote_waitForLogMessage(t, otbrSnap, "Start Thread Border Agent: OK", start)
+	t.Log("OTBR on remote device is ready")
+}
+
+func remote_deployGPIOCommander(t *testing.T) {
+
+	t.Cleanup(func() {
+		remote_exec(t, "sudo snap remove --purge "+matterGPIOSnap)
+	})
+
+	installCommand := fmt.Sprintf("sudo snap install %s --edge", matterGPIOSnap)
+	extraInterface := ""
+	if remoteSnapPath != "" {
+		installCommand = fmt.Sprintf("sudo snap install --dangerous %s", remoteSnapPath)
+		extraInterface = "sudo snap connect " + matterGPIOSnap + ":custom-gpio " + matterGPIOSnap + ":custom-gpio-dev"
+	}
+
+	start := time.Now().UTC()
+
+	commands := []string{
+		"sudo snap remove --purge " + matterGPIOSnap,
+		installCommand,
+		extraInterface,
+		"sudo snap set" + matterGPIOSnap + "args=\"--thread\"",
+		"sudo snap set" + matterGPIOSnap + "gpio=\"16\"",
+		"sudo snap connect " + matterGPIOSnap + ":avahi-control",
+		"sudo snap connect " + matterGPIOSnap + ":otbr-dbus-wpan0 " + otbrSnap + ":dbus-wpan0",
+		"sudo snap start " + matterGPIOSnap,
```

Comment:
Since this application is the subject of these tests, it is important to capture its logs on test failures. Otherwise, it will be hard to identify the startup error.

I had to run `sudo journalctl -f  | grep commander` on the remote device to capture https://github.com/canonical/matter-pi-gpio-commander/pull/68/files#r1589256451